### PR TITLE
feat(MessageEvents): Promisable send/edit listeners

### PR DESCRIPTION
**Pull request by hunt-g**
makes pre-send and pre-edit events optionally asynchronous

### CODE REVIEW

File: src/api/MessageEvents.ts [hunk 1]
```diff
@@ -19,6 +19,7 @@
 import Logger from "@utils/Logger";
 import { MessageStore } from "@webpack/common";
 import type { Channel, Message } from "discord-types/general";
+import type { Promisable } from "type-fest";
```
Comment (Nuckyz) [inline src/api/MessageEvents.ts:22]:
```suggestion
```

File: src/api/MessageEvents.ts [hunk 2]
```diff
@@ -41,16 +42,16 @@ export interface MessageExtra {
     stickerIds?: string[];
 }
 
-export type SendListener = (channelId: string, messageObj: MessageObject, extra: MessageExtra) => void | { cancel: boolean; };
-export type EditListener = (channelId: string, messageId: string, messageObj: MessageObject) => void;
+export type SendListener = (channelId: string, messageObj: MessageObject, extra: MessageExtra) => Promisable<void | { cancel: boolean; }>;
```
Comment (Nuckyz) [inline src/api/MessageEvents.ts:45]:
```suggestion
type SendListernerReturn = void | { cancel: boolean; };

export type SendListener = (channelId: string, messageObj: MessageObject, extra: MessageExtra) => SendListernerReturn | Promise<SendListernerReturn>;
```

File: src/api/MessageEvents.ts [hunk 3]
```diff
@@ -41,16 +42,16 @@ export interface MessageExtra {
...
+export type EditListener = (channelId: string, messageId: string, messageObj: MessageObject) => Promisable<void>;
```
Comment (Nuckyz) [inline src/api/MessageEvents.ts:46]:
```suggestion
export type EditListener = (channelId: string, messageId: string, messageObj: MessageObject) => void | Promise<void>;
```

File: src/plugins/apiMessageEvents.ts [hunk 4]
```diff
@@ -22,16 +22,16 @@ import definePlugin from "@utils/types";
 export default definePlugin({
     name: "MessageEventsAPI",
     description: "Api required by anything using message events.",
-    authors: [Devs.Arjix],
+    authors: [Devs.Arjix, Devs.hunt],
     patches: [
         {
             find: "sendMessage:function",
```
Comment (Nuckyz) [inline src/plugins/apiMessageEvents.ts:28]:
```suggestion
            find: '"MessageActionCreators"',
```

File: src/plugins/apiMessageEvents.ts [hunk 5]
```diff
@@ -22,16 +22,16 @@ import definePlugin from "@utils/types";
...
             replacement: [{
-                match: /(?<=_sendMessage:function\([^)]+\)){/,
-                replace: "{if(Vencord.Api.MessageEvents._handlePreSend(...arguments)){return;};"
+                match: /(_sendMessage:function.*?{)(.*?)(},[^}]*?:function)/,
```
Comment (Nuckyz) [inline src/plugins/apiMessageEvents.ts:30]:
```suggestion
                match: /(_sendMessage:function\((?:\i,?).+?\){)(.+?)(?=},.{1,30}:function)/,
```

File: src/plugins/apiMessageEvents.ts [hunk 6]
```diff
@@ -22,16 +22,16 @@ import definePlugin from "@utils/types";
...
+                replace: "$1return Vencord.Api.MessageEvents._handlePreSend(...arguments).then(()=>{$2})$3"
```
Comment (Nuckyz) [inline src/plugins/apiMessageEvents.ts:31]:
```suggestion
                replace: "$1return Vencord.Api.MessageEvents._handlePreSend(...arguments).then(()=>{$2})"
```

File: src/plugins/apiMessageEvents.ts [hunk 7]
```diff
@@ -22,16 +22,16 @@ import definePlugin from "@utils/types";
...
             }, {
-                match: /(?<=\beditMessage:function\([^)]+\)){/,
-                replace: "{Vencord.Api.MessageEvents._handlePreEdit(...arguments);"
+                match: /(editMessage:function.*?{)(.*?)(},[^}]*?:function)/,
```
Comment (Nuckyz) [inline src/plugins/apiMessageEvents.ts:33]:
```suggestion
                match: /(editMessage:function\((?:\i,?).+?\){)(.+?)(?=},.{1,30}:function)/,
```

File: src/plugins/apiMessageEvents.ts [hunk 8]
```diff
@@ -22,16 +22,16 @@ import definePlugin from "@utils/types";
...
+                replace: "$1return Vencord.Api.MessageEvents._handlePreEdit(...arguments).then(()=>{$2})$3"
```
Comment (Nuckyz) [inline src/plugins/apiMessageEvents.ts:34]:
```suggestion
                replace: "$1return Vencord.Api.MessageEvents._handlePreEdit(...arguments).then(()=>{$2})"
```

**Review by Nuckyz — CHANGES_REQUESTED** [review]

**Comment by Vendicated** [comment]
I simplified and fixed the now broken patches, thank you and sorry for the wait!

**Pull request merged**